FP-Quant support

BlackSamorez opened the pull request:
# This PR adds support for the FP-Quant method.

The goal of this PR is to integrate inference and training support for the FP-Quant method that utilizes the Hadamard Transform for efficient weights+activations quantization. When using it with MXFP4 and MSE-based scaling, it implements [Quartet forward pass](https://arxiv.org/abs/2505.14669). We're also working on adding NVFP4 support and backward pass support.

Currently, we're working on the kernels [here](https://github.com/IST-DASLab/qutlass), and the integration [here](https://github.com/IST-DASLab/FP-Quant).

Installation:
 1. Install `qutlass`: `git clone https://github.com/IST-DASLab/qutlass.git && cd qutlass && pip install --no-build-isolation .`
 2. Install `fp_quant`: `pip install fp_quant`

Usage:
 1. Use as JIT quantization from any BF16 model by passing `quantization_config=FPQuantConfig()`
 2. Calibrate with GPTQ with [the repo](https://github.com/IST-DASLab/FP-Quant) with `--real_quant`.
 3. Use pre-quantized models from hub: coming soon...

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker
- vision models: @amyeroberts, @qubvel
- speech models: @eustlb
- graph models: @clefourrier

Library:

- flax: @gante and @Rocketknight1
- generate: @zucchini-nlp (visual-language models) or @gante (all others)
- pipelines: @Rocketknight1
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @zach-huggingface and @SunMarc
- chat templates: @Rocketknight1

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc @zach-huggingface
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc @MekkCyber

Documentation: @stevhliu

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @Rocketknight1
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

BlackSamorez committed:
quartet

MekkCyber commented on src/transformers/quantizers/quantizer_quartet_qat.py:
```diff
@@ -0,0 +1,164 @@
+
+    def update_torch_dtype(self, torch_dtype: "torch.dtype") -> "torch.dtype":
+        if torch_dtype is None:
+            logger.info("`torch_dtype` is None. Setting `torch_dtype=torch.float16` for FLUTE compatibility.")
+            torch_dtype = torch.float16
+        elif torch_dtype != torch.float16 and torch_dtype != torch.bfloat16:
+            raise ValueError(
+                f"Invalid `torch_dtype` {torch_dtype}. HIGGS quantization only supports `torch_dtype=torch.float16` or `torch_dtype=torch.bfloat16`."
+            )
```
same some typos from the HIGGS integration

MekkCyber commented on src/transformers/utils/import_utils.py:
```diff
@@ -1149,6 +1152,18 @@ def is_quark_available():
+    return _quartet_qat_available
+
+
+def is_qutlass_available():
+    return _qutlass_available
+
```
I can't find a distribution for `qutlass`, is it not released yet ?

BlackSamorez replied:
It has just been released: https://github.com/IST-DASLab/qutlass

MekkCyber commented on src/transformers/quantizers/quantizer_quartet_qat.py:
```diff
@@ -0,0 +1,164 @@
+        quartet_qat_modules = {
+            name: module for name, module in model.named_modules() if isinstance(module, QuartetLinear)
+        }
+        for name, module in tqdm(quartet_qat_modules.items(), desc="Pre-processing Quartet QAT modules", leave=False):
+            pass
+            # module.pre_forward()
+
```
What’s meant to happen here exactly ?

MekkCyber commented on src/transformers/quantizers/quantizer_quartet_qat.py:
```diff
@@ -0,0 +1,164 @@
+        from quartet_qat import QuartetLinear
+
+        module, tensor_name = get_module_from_name(model, param_name)
+        if isinstance(module, QuartetLinear) and tensor_name == "weight":
+            # Only quantize weights of QuartetLinear modules that are not already quantized
+            return True
+        else:
+            return False
```
is the bias quantized too ?

MekkCyber commented on src/transformers/quantizers/quantizer_quartet_qat.py:
```diff
@@ -0,0 +1,164 @@
+        assert isinstance(module, QuartetLinear), f"Module {param_name} is not a QuartetLinear somehow..."
+
+        module.weight = torch.nn.Parameter(param_value.to(target_device), requires_grad=module.weight.requires_grad)
+        module.pre_forward()
+
```
what's happening here ?

MekkCyber reviewed (commented):
Hi @BlackSamorez ! Thanks a lot for this addition 🤗 ! Left a few comments !

BlackSamorez committed:
quartet qat -> quartet

BlackSamorez committed:
format

BlackSamorez committed:
bf16 backward

BlackSamorez committed:
interfaces

BlackSamorez committed:
forward_method

kooshi commented:
Hi @BlackSamorez, I'm really looking forward to experimenting with this.

When can we expect to have the kernels public so we can begin testing, even if they are still WIP?

BlackSamorez committed:
quartet -> fp_quant

BlackSamorez committed:
style

BlackSamorez committed:
Merge branch 'huggingface:main' into main

BlackSamorez committed:
List -> list

BlackSamorez committed:
list typing

BlackSamorez committed:
Merge branch 'main' of github.com:BlackSamorez/transformers

BlackSamorez committed:
fixed format and annotations

BlackSamorez committed:
test_fp_quant

BlackSamorez committed:
docstrings and default dtypes

BlackSamorez commented:
@MekkCyber Hi, thanks for reviewing this!
It took us a while, but all the kernels necessary for inference have been published: I've updated the PR description.
May I ask you to do another pass? Your previous comments mostly don't apply anymore because of refactoring.

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -1550,6 +1551,63 @@ def post_init(self):
+            The method to use for the forward pass.
+        backward_dtype (`str`, *optional*, defaults to `"bf16"`):
+            The dtype to use for the backward pass.
+        store_master_weights (`bool`, *optional*, defaults to `False`):
+            Whether to store the master weights.
```
in which context storing master weights could be useful ?

BlackSamorez replied:
In the context of QAT, which we'll add in a later release: we're still working on the quantized backward pass kernels. But I thought it would make sense to include this option right away to not have to edit the config later.

BlackSamorez replied:
Added this to docstring

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -1550,6 +1551,63 @@ def post_init(self):
+    Args:
+        forward_dtype (`str`, *optional*, defaults to `"mxfp4"`):
+            The dtype to use for the forward pass.
+        forward_method (`str`, *optional*, defaults to `"abs_max"`):
+            The method to use for the forward pass.
```
we have `absmax` and `quest` for this arg. can you explain a bit what quest does ?

BlackSamorez replied:
Added docstring explanation

SunMarc commented on src/transformers/utils/quantization_config.py:
```diff
@@ -1550,6 +1551,63 @@ def post_init(self):
+            The dtype to use for the backward pass.
+        store_master_weights (`bool`, *optional*, defaults to `False`):
+            Whether to store the master weights.
+        hadamard_group_size (`int`, *optional*, defaults to 32):
+            The group size for the hadamard transform.
```
explain a bit what this does

BlackSamorez replied:
Improved docstring

SunMarc commented on tests/quantization/fp_quant_integration/test_fp_quant.py:
```diff
@@ -0,0 +1,147 @@
+from transformers.utils import is_accelerate_available, is_torch_available
+
+
+if is_torch_available():
+    pass
+
+if is_accelerate_available():
+    pass
```
remove

SunMarc commented on src/transformers/quantizers/quantizer_fp_quant.py:
```diff
@@ -0,0 +1,168 @@
+            return
+
+        module.weight = torch.nn.Parameter(param_value.to(target_device))
+        module.pre_forward()
```
really nice to put all the quantization logic there

SunMarc reviewed (commented):
Thanks for iterating on the PR and congrats on the release ! The only major thing missing before we merge this is some documentation for this new method ! Please ping me when it's done and I'll merge the PR !

SunMarc committed:
Merge branch 'main' into main

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_38696). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BlackSamorez committed:
better docstring and removed noop checks

BlackSamorez committed:
docs

BlackSamorez commented:
@SunMarc added docs, improved docstring, cleaned the code where you asked.

BlackSamorez committed:
pseudoquantization support to test on non-blackwell

SunMarc reviewed (approved):
LGTM ! Just a minor nit, let me know when you want to merge this

BlackSamorez commented:
Actually, give me a minute. I'm adding Triton pseudo-quantization kernels for people without Blackwell GPUs to be able to evaluate the models (although without speedups).

BlackSamorez committed:
pseudoquant

BlackSamorez committed:
Pseudoquant docs

BlackSamorez commented:
Added pseudoquantization, updated requirements to run the method with it (doesn't require `qutlass` like that). Added pseudoquant tests. Updated the documentation.

BlackSamorez committed:
Update docs/source/en/quantization/fp_quant.md

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

BlackSamorez commented on docs/source/en/quantization/fp_quant.md:
```diff
@@ -0,0 +1,66 @@
+
+model = AutoModelForCausalLM.from_pretrained(
+    "qwen/Qwen3-8B",
+    quantization_config=HiggsConfig(bits=4),
```
```suggestion
    quantization_config=FPQuantConfig(),
```

BlackSamorez committed:
Update docs/source/en/quantization/fp_quant.md

BlackSamorez commented on docs/source/en/quantization/fp_quant.md:
````diff
@@ -0,0 +1,66 @@
+
+```python
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer, HiggsConfig
````
```suggestion
from transformers import AutoModelForCausalLM, AutoTokenizer, FPQuantConfig
```

BlackSamorez committed:
Update docs/source/en/quantization/fp_quant.md

MekkCyber commented on docs/source/en/quantization/fp_quant.md:
```diff
@@ -0,0 +1,66 @@
+
+# FP-Quant
+
+[FP-Quant](https://github.com/IST-DASLab/FP-Quant) is a family of quantization algorithms tailored for the Blackwell generation of Nvidia GPUs. The goal is to allow for efficient post-training quantization (PTQ) and quantization-aware trainin (QAT) of LLMs in the MXFP4 and NVFP4 data-types.
```
if you can add a link to some doc about `mxfp4` and `nvfp4` that would be great 🤗

MekkCyber commented on src/transformers/quantizers/quantizer_fp_quant.py:
```diff
@@ -0,0 +1,190 @@
+        )
+        model.config.quantization_config = self.quantization_config
+
+    def _process_model_after_weight_loading(self, model: "PreTrainedModel", **kwargs):
+        from fp_quant import FPQuantLinear
+
+        fp_quant_modules = {
+            name: module for name, module in model.named_modules() if isinstance(module, FPQuantLinear)
+        }
+        for name, module in fp_quant_modules.items():
+            if not self.quantization_config.store_master_weights and module.weight is not None:
+                module.weight = None
```
why are we doing this here when we can do it in `create_quantized_param` ? it adds two loops over modules

BlackSamorez replied:
I deleted it and it still works. It was probably needed for the previous dispatch implementation I had. I removed it now.

MekkCyber commented on src/transformers/utils/quantization_config.py:
```diff
@@ -1550,6 +1551,67 @@ def post_init(self):
+        r"""
+        Safety checker that arguments are correct - also replaces some NoneType arguments with their default values.
+        """
+        if self.forward_dtype not in ["mxfp4"]:
+            raise ValueError("forward_dtype must be mxfp4 for now")
+        if self.forward_method not in ["abs_max", "quest"]:
+            raise ValueError("forward_method must be abs_max or quest for now")
+        if self.backward_dtype not in ["bf16"]:
+            raise ValueError("backward_dtype must be bf16 for now")
+        if self.hadamard_group_size not in [32]:
+            raise ValueError("hadamard_group_size must be 32 for now")
```
small nit
```suggestion
        if self.forward_dtype not in ["mxfp4"]:
            raise ValueError("Only 'mxfp4' is supported for forward_dtype for now.")
        if self.forward_method not in ["abs_max", "quest"]:
            raise ValueError("Only 'abs_max' and 'quest' are supported for forward_method for now.")
        if self.backward_dtype not in ["bf16"]:
            raise ValueError("Only 'bf16' is supported for backward_dtype for now.")
        if self.hadamard_group_size not in [32]:
            raise ValueError("Only a hadamard_group_size of 32 is supported for now.")
```

MekkCyber commented on tests/quantization/fp_quant_integration/test_fp_quant.py:
```diff
@@ -0,0 +1,231 @@
+        Setup quantized model
+        """
+        quantization_config = FPQuantConfig(pseudoquantization=True)
+        cls.tokenizer = AutoTokenizer.from_pretrained(cls.model_name)
+        cls.quantized_model = AutoModelForCausalLM.from_pretrained(
+            cls.model_name, device_map=cls.device_map, quantization_config=quantization_config, low_cpu_mem_usage=True
+        )
```
defaults to True anyway
```suggestion
        cls.tokenizer = AutoTokenizer.from_pretrained(cls.model_name)
        cls.quantized_model = AutoModelForCausalLM.from_pretrained(
            cls.model_name, device_map=cls.device_map, quantization_config=quantization_config
        )
```

MekkCyber commented on tests/quantization/fp_quant_integration/test_fp_quant.py:
```diff
@@ -0,0 +1,231 @@
+            output = model.generate(**input_ids, max_new_tokens=self.max_new_tokens)
+            self.assertEqual(self.tokenizer.decode(output[0], skip_special_tokens=True), self.EXPECTED_OUTPUT)
+
+    @require_torch_multi_gpu
+    def test_quantized_model_multi_gpu(self):
```
shouldn't we add here an other requirement for compute cap >= 10.0 ?

MekkCyber commented on docs/source/en/quantization/fp_quant.md:
```diff
@@ -0,0 +1,66 @@
+
+A **Blackwell-generation GPU is required** to run the kernels. Runtime support for FP-Quant is implemented through the [QuTLASS](https://github.com/IST-DASLab/qutlass) library and a lightweight PyTorch interface lib [`fp_quant`](https://github.com/IST-DASLab/FP-Quant/tree/master/inference_lib). We recommend installing the former **from source** and the latter with  `pip install fp_quant`.
+
+Users **without a Blackwell-generation GPU** , can use the method with `quantization_config=FPQuantConfig(pseudoquant=True)` without having to install [QuTLASS](https://github.com/IST-DASLab/qutlass). This would provide no speedups but would fully emulate the effect of quantization.
```
Does pseudo quantization work on any recent gpu ? or it requires >=9.0

MekkCyber replied:
Sounds good thanks!

MekkCyber reviewed (commented):
Thanks for this great addition @BlackSamorez ! lefts minor comments and questions

BlackSamorez committed:
Update src/transformers/utils/quantization_config.py

Co-authored-by: Mohamed Mekkouri <93391238+MekkCyber@users.noreply.github.com>

BlackSamorez committed:
Update tests/quantization/fp_quant_integration/test_fp_quant.py

Co-authored-by: Mohamed Mekkouri <93391238+MekkCyber@users.noreply.github.com>

SunMarc reviewed (commented):
Thanks, just a few nits and we are good to merge

BlackSamorez committed:
Update tests/quantization/fp_quant_integration/test_fp_quant.py

Co-authored-by: Marc Sun <57196510+SunMarc@users.noreply.github.com>

BlackSamorez committed:
small test fixes

BlackSamorez committed:
dockerfile update

BlackSamorez committed:
Merge branch 'main' of github.com:BlackSamorez/transformers

BlackSamorez committed:
spec link

BlackSamorez committed:
removed `_process_model_after_weight_loading`

SunMarc committed:
Merge branch 'main' into main

SunMarc reviewed (approved):
LGTM ! Thanks for iterating !

BlackSamorez committed:
toctree

BlackSamorez committed:
Merge branch 'main' of github.com:BlackSamorez/transformers

BlackSamorez commented:
@SunMarc it hit job cancellation somehow. Might need a restart. It should be good.

SunMarc committed:
Merge branch 'main' into main

SunMarc commented:
Merged ! Thanks for your work

SunMarc commented:
Hey @BlackSamorez, is there a way to make fp_quant compatible with py3.9 ? Our CI runs on this version but fp_quant requires 3.11

BlackSamorez commented:
I guess I'll have to remove match-case constructions and it'll work.
Why run on 3.9 in 2025 though?

BlackSamorez commented:
I guess I'll have to remove match-case constructions and it'll work.
Why run on 3.9 in 2025 though?
